### ISSUE TITLE
Out of memory on get_margeff() 

### ISSUE THREAD

**Issue by dpavlic** [statsmodels/statsmodels#4158]
I have a dataset of about 400 thousand observations. It is not too large, at around 100 MB of space. Version is 0.8.0. Unfortunately, as data is confidential, I cannot include it, but I include the code below.

I estimate the model with:

```
import statsmodels.formula.api as st
import pandas as pd

d = pd.read_feather('mydata.feather')
m = st.mnlogit(
    formula = 'y ~ sex + vismin + special_needs + esl + t_score1 + t_score2 + t_score3', data = d
).fit()
```

So far everything works as expected. The mlogit estimates look fine. However, the following:

```
m.get_margeff()
```

Never completes, and runs out of memory (it takes 30 GB on a 32 GB machine before it fails). There is an object in there that appears to just keep growing.

Getting marginal effects at the means works fine:

```
m.get_margeff(at = 'mean')
```

Any hints would be very helpful as to where the trouble might be. For comparison, Stata does this task with about 1 1/2 GB memory footprint throughout.

**Comment by josef-pkt** [comment]
What's the shape of your params? ie. how many levels of y and how many columns in exog?

I'm not really familiar with all the details of the margin computation. The problem might be that it is completely vectorized so that some intermediate, most likely derivative, matrices become very large.
But I have no guess which step, function or method of the computation uses very large arrays.

Unless there is some computation that wastes memory, the only option I can think of is to add some loops so the intermediate results are computed in batches.

(One other possibility, which might not be likely, is that something prevents fast garbage collection during some computations. In that case older intermediate results might still be in memory which can increase memory very fast with large numpy arrays.)

**Comment by dpavlic** [comment]
There's 4 categories in y. In actual fact, there's 9 independent variables are all categorical, 6 are binary categoricals, the other 3 are categoricals with 7 categories. The issue seems to occur when the 3 with 7 categories are added to the regression. I tried doing just this to a variable with 7 categories:

```
margins = m.get_margeff(atexog = 't_score1')
```

I might be misusing that, but my assumption is the margeff gets calculated only for that particular variable, the idea being to loop through all the variables.

However: even here, the memory goes over 28 GB for just that one variable, and I eventually get an out of memory error.

**Comment by josef-pkt** [comment]
So if I understand you correctly, then params is 3 by (6+3*7) makes 81 parameter time 400 thousand obs makes over 32 million elements in the derivative matrix for all observations.

You could try to loop in batches over exog. I think the "overall" margeff is additive, but I'm not sure I remember correctly or checked the math for this case.

e.g. try margeff but with just the first 10000 rows of your exog, and check how much peak memory that takes and if memory goes back down when it's finished.

It would be good to find a temporary workaround if possible.
The proper fix will take more time. We need an artificial example to reproduce the memory consumption, and then work through the entire code path. Given that the margeff code is not very simple (i.e. it has a lot of moving parts and helper functions) it will take more time to do that. (Time which I won't have right now.)

> margins = m.get_margeff(atexog = 't_score1')

I'm not sure that works, `atexog` should be a dict or ndarray or is ignored otherwise, AFAICS in function `discrete_margins._get_margeff_exog`
but this just fixes the values of one exog column and doesn't change that all margeffs are computed
(eg. what's the average if everyone where female and otherwise unchanged) IIUC/IIRC

**Comment by dpavlic** [comment]
I will try to see if I can find something similarly complicated and in the public domain to check against. In the mean time, I'd like to try a workaround.  In particular, I'm not sure I understand this part:

> You could try to loop in batches over exog.

How precisely would I go around doing that? Of course, I can estimate the model with a 10,000 obs sample, then call get_margeff and see if that works: it does -- and it takes about 500 MB memory which is released. But I assume you meant something else?

Thanks for all your help.

**Comment by josef-pkt** [comment]
I meant something different.

Use your original model, estimated with the full data, but use the `exog` argument in margeff

```
exog = model.exog
i = 0
me = m.get_margeff(exog=exog[i*1000 : (i+1) * 10000])
```

If that works, then loop over i to get the entire exog, and then take weighted average (weighted  by batch size if the last batch is smaller).
And if that looks reasonable, then we need to figure out standard errors and similar.

(I haven't tried this out, or don't remember if I did, so not everything might work how I think it should work.)

**Comment by dpavlic** [comment]
Got it. I'll see what that gets me, thanks.

**Comment by dpavlic** [comment]
Hmm, this produced a different problem:

```
      1 i = 0
      2 exog = model.exog
----> 3 me = m.get_margeff(exog=exog[i * 1000 : (i+1) * 10000])

TypeError: get_margeff() got an unexpected keyword argument 'exog'
```

**Comment by josef-pkt** [comment]
Sorry, I got confused about arguments in helper functions

it should be `atexog`: `me = m.get_margeff(atexog=exog[i * 1000 : (i+1) * 10000])`
the atexog argument needs to be a numpy array. There is an `isinstance(atexog, np.ndarray)` check in the helper function `_get_margeff_exog`.

**Comment by dpavlic** [comment]
I ran this, and the marginal effects do not match; however, I'm comparing with Stata's marginal effect results. 

I am going to take a random sample of maybe a quarter of all observations, and see if I can get marginal effects to line up  between 'batch' mode and estimating it all in one go.

**Comment by josef-pkt** [comment]
Note, by default statsmodels margeff assume a continuous exog. I don't remember what Stata uses as default and whether it matches up. There is an option for binary/dummy variable treatement.

**Comment by dpavlic** [comment]
To update, the AMEs are however matched up internally. I just ran a subsample of 100,000 observations and also a 'batch' job estimating 1/4 of the sample * 4, and the average returns the same AMEs as the full 100,000 sample. I'll check the dummy treatment option, but it seems like this approach does work.

**Comment by josef-pkt** [comment]
@dpavlic Did the results with dummy treatment match Stata?

**Comment by dpavlic** [comment]
No... but this is because running dummy = True actually produced an error. I will be able to get the stack trace soon, but I'm afraid I'm battling a cocktail of viruses at the same time so it will probably need to wait a few days.

**Comment by dpavlic** [comment]
Sorry for the delay, life intervened this past bit. Here is the traceback:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-36-a89bf7db787a> in <module>()
----> 1 me1 = mf.get_margeff(atexog=exog[0:100666], dummy = True)

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py in get_margeff(self, at, method, atexog, dummy, count)
   2502         """
   2503         from statsmodels.discrete.discrete_margins import DiscreteMargins
-> 2504         return DiscreteMargins(self, (at, method, atexog, dummy, count))
   2505 
   2506     def summary(self, yname=None, xname=None, title=None, alpha=.05,

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_margins.py in __init__(self, results, args, kwargs)
    423         self._cache = resettable_cache()
    424         self.results = results
--> 425         self.get_margeff(*args, **kwargs)
    426 
    427     def _reset(self):

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_margins.py in get_margeff(self, at, method, atexog, dummy, count)
    704                                                 model._derivative_exog,
    705                                                 dummy_idx, count_idx,
--> 706                                                 method, J)
    707 
    708             # reshape for multi-equation

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_margins.py in margeff_cov_with_se(model, params, exog, cov_params, at, derivative, dummy_ind, count_ind, method, J)
    342     cov_me = margeff_cov_params(model, params, exog, cov_params, at,
    343                                               derivative, dummy_ind,
--> 344                                               count_ind, method, J)
    345     return cov_me, np.sqrt(np.diag(cov_me))
    346 

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_margins.py in margeff_cov_params(model, params, exog, cov_params, at, derivative, dummy_ind, count_ind, method, J)
    322         if dummy_ind is not None:
    323             jacobian_mat = _margeff_cov_params_dummy(model, jacobian_mat,
--> 324                                 params, exog, dummy_ind, method, J)
    325         if count_ind is not None:
    326             jacobian_mat = _margeff_cov_params_count(model, jacobian_mat,

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_margins.py in _margeff_cov_params_dummy(model, cov_margins, params, exog, dummy_ind, method, J)
    212         exog0[:,i] = 0
    213         exog1[:,i] = 1
--> 214         dfdb0 = model._derivative_predict(params, exog0, method)
    215         dfdb1 = model._derivative_predict(params, exog1, method)
    216         dfdb = (dfdb1 - dfdb0)

~/.conda/envs/myenv/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py in _derivative_predict(self, params, exog, transform)
    635         X = np.tile(exog, J-1)
    636         # this is the derivative wrt the base level
--> 637         F0 = -repeat_eXB * X / sum_eXB ** 2
    638         # this is the derivative wrt the other levels when
    639         # dF_j / dParams_j (ie., own equation)

ValueError: operands could not be broadcast together with shapes (100666,12) (100666,75) 
```

**Comment by josef-pkt** [comment]
another case of MemoryError in margeff is in #5247
